controllers/glanceapi_controller.go

```diff
@@ -288,61 +286,90 @@ func (r *GlanceAPIReconciler) reconcileInit(
 	r.Log.Info(fmt.Sprintf("Reconciling Service '%s' init", instance.Name))
 
 	//
-	// expose the service (create service, route and return the created endpoint URLs)
+	// create service/s
 	//
-	ports := map[endpoint.Endpoint]endpoint.Data{}
+	glanceEndpoints := map[service.Endpoint]endpoint.Data{}
 	if instance.Spec.APIType == glancev1.APIInternal {
-		ports[endpoint.EndpointInternal] = endpoint.Data{
+		glanceEndpoints[service.EndpointInternal] = endpoint.Data{
 			Port: glance.GlanceInternalPort,
 		}
 	} else {
-		ports[endpoint.EndpointPublic] = endpoint.Data{
+		glanceEndpoints[service.EndpointPublic] = endpoint.Data{
 			Port: glance.GlancePublicPort,
 		}
 	}
+	apiEndpoints := make(map[string]string)
 
-	for _, metallbcfg := range instance.Spec.ExternalEndpoints {
-		portCfg := ports[metallbcfg.Endpoint]
-		portCfg.MetalLB = &endpoint.MetalLBData{
-			IPAddressPool:   metallbcfg.IPAddressPool,
-			SharedIP:        metallbcfg.SharedIP,
-			SharedIPKey:     metallbcfg.SharedIPKey,
-			LoadBalancerIPs: metallbcfg.LoadBalancerIPs,
+	for endpointType, data := range glanceEndpoints {
+		endpointName := glance.ServiceName + "-" + string(endpointType)
+		svcOverride := instance.Spec.Override.Service
+
+		exportLabels := util.MergeStringMaps(
+			serviceLabels,
+			map[string]string{
+				string(endpointType): "true",
+			},
+		)
+
+		// Create the service
+		svc, err := service.NewService(
+			service.GenericService(&service.GenericServiceDetails{
```

Comment:
> IIUC the `NewService` function will `patch` the `spec` for the generated services [1], then we build the actual `svc` on L341. The service operator is still responsible to build services associated with the Deployment. What's new in the service operator is that we're not going to "Expose the endpoints" (aka: create Routes) anymore, but only:
> 
> * generate a service associated with the Deployment
> * Register endpoints in keystone

yes, thats correct. The service operator creates the service based on the overrides it gets from either openstack-operator or if you created it standalone from a user. When a route got created you pass in the route base url (http://foo.bar) and the service operator uses this information for register the keystone endpoint.

> 
> This means we're not actually catching in this layer any issue associated with the Route creation, which can be seen here [2]

I am not sure I got that. the route gets created before the service from the openstack-operator. so if there is an issue it will not create the service and condition gets set on the openstack operator.

> 
> [1] https://github.com/openstack-k8s-operators/lib-common/pull/313/files#diff-ab24654840bf505a081fc14bb7a6283ef70c4fdc5e57896971af03c5b8c2d1aeR54 [2] https://github.com/openstack-k8s-operators/openstack-operator/pull/416/files#diff-dc49c70e72e8b711bfefebe68ede374400ee1c2c0fe24c4a25a1a6c352a1a489R132

Reply:
> > IIUC the `NewService` function will `patch` the `spec` for the generated services [1], then we build the actual `svc` on L341. The service operator is still responsible to build services associated with the Deployment. What's new in the service operator is that we're not going to "Expose the endpoints" (aka: create Routes) anymore, but only:
> > 
> > * generate a service associated with the Deployment
> > * Register endpoints in keystone
> 
> yes, thats correct. The service operator creates the service based on the overrides it gets from either openstack-operator or if you created it standalone from a user. When a route got created you pass in the route base url (http://foo.bar) and the service operator uses this information for register the keystone endpoint.
ack, thanks for clarifying this.

> 
> > This means we're not actually catching in this layer any issue associated with the Route creation, which can be seen here [2]
> 
> I am not sure I got that. the route gets created before the service from the openstack-operator. so if there is an issue it will not create the service and condition gets set on the openstack operator.

Now I think I have a better understanding about the flow and the key fact is the `Route`  that is created before the service:

1. the OpenStack operator -> `CreateRouteAndServiceOverride`
2. serviceOverrides are produced by the previous step `serviceOverrides[string(endpointType)] = *svcOverride` and passed to the Service definition for each endpoint type  (e.g., `glance.Spec.GlanceAPIExternal.Override.Service = ptr.To(serviceOverrides[string(service.EndpointPublic)])`.

> 
> > [1] https://github.com/openstack-k8s-operators/lib-common/pull/313/files#diff-ab24654840bf505a081fc14bb7a6283ef70c4fdc5e57896971af03c5b8c2d1aeR54 [2] https://github.com/openstack-k8s-operators/openstack-operator/pull/416/files#diff-dc49c70e72e8b711bfefebe68ede374400ee1c2c0fe24c4a25a1a6c352a1a489R132